### PR TITLE
Implement a horizontal panel for the workbench

### PR DESCRIPTION
**Pull request by isidorn**
Adding horizontal panel to the workbench (see #1875).

### CODE REVIEW

File: src/vs/workbench/browser/parts/activitybar/activitybarPart.ts [hunk 1]
```diff
@@ -61,28 +61,28 @@ export class ActivitybarPart extends Part implements IActivityService {
 	private registerListeners(): void {
 
 		// Activate viewlet action on opening of a viewlet
-		this.toUnbind.push(this.eventService.addListener(EventType.VIEWLET_OPENING, (e: ViewletEvent) => this.onViewletOpening(e)));
+		this.toUnbind.push(this.eventService.addListener(EventType.VIEWLET_OPENING, (e: CompositeEvent) => this.onViewletOpening(e)));
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]: @isidorn I think we can generalize this event to be COMPOSITE_OPENING and onCompositeOpening() because the listener seems to use the ID of the composite to do something and it also seems to be the only client through out the workbench.

File: src/vs/workbench/browser/parts/compositePart.ts [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+		let createCompositePromise: TPromise<void>;
+
+		// Composits created for the first time
+		let compositeContainer = this.mapCompositeToCompositeContainer[composite.getId()];
+		if (!compositeContainer) {
+
+			// Build Container off-DOM
+			compositeContainer = $().div({
+				'class': 'viewlet',
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/compositePart.ts]: @isidorn this should probably be passed in from the outside so that we can tweak the CSS for viewlets and panels separately.

File: src/vs/workbench/browser/parts/compositePart.ts [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+		// Report progress for slow loading composits (but only if we did not create the composits before already)
+		let progressService = this.mapProgressServiceToComposite[composite.getId()];
+		if (progressService && !compositeContainer) {
+			this.mapProgressServiceToComposite[composite.getId()].showWhile(createCompositePromise, this.partService.isCreated() ? 800 : 3200 /* less ugly initial startup */);
+		}
+
+		// Fill Content and Actions
+		return createCompositePromise.then(() => {
+
+			// Make sure that the user meanwhile did not open another composite or closed the sidebar
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/compositePart.ts]: @isidorn sidebar reference in comment

File: src/vs/workbench/browser/parts/compositePart.ts [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+			this.telemetryActionsListener = this.toolBar.actionRunner.addListener2(events.EventType.RUN, (e: any) => {
+
+				// Check for Error
+				if (e.error && !errors.isPromiseCanceledError(e.error)) {
+					this.messageService.show(Severity.Error, e.error);
+				}
+
+				// Log in telemetry
+				if (this.telemetryService) {
+					this.telemetryService.publicLog('workbenchActionExecuted', { id: e.action.id, from: 'sideBar' });
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/compositePart.ts]: @isidorn the "from: sideBar" value should be passed in from the subclass of the part

File: src/vs/workbench/browser/parts/compositePart.ts [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+				}
+			});
+
+			// Indicate to composite that it is now visible
+			return composite.setVisible(true).then(() => {
+
+				// Make sure that the user meanwhile did not open another composite or closed the sidebar
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/compositePart.ts]: @isidorn "sidebar" being mentioned in comment

File: src/vs/workbench/browser/parts/activitybar/activitybarPart.ts (hunk 1 above)
Comment (bpasero) [inline src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]: I see why this happens because opening a viewlet and panel is not yet shared in the compositePart, I will comment below about it.

File: src/vs/workbench/browser/parts/panel/panelPart.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		partService: IPartService,
+		keybindingService: IKeybindingService,
+		id: string
+	) {
+		super(messageService, storageService, eventService, telemetryService, contextMenuService, partService, keybindingService,
+			(<PanelRegistry>Registry.as(PanelExtensions.Panels)), PanelPart.activePanelSettingsKey, id);
+	}
+
+	public openPanel(id: string, focus?: boolean): TPromise<Panel> {
+		if (this.blockOpeningPanel) {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/panel/panelPart.ts]: @isidorn I would try to push this down to the compositePart because this seems like lots of duplicated code also in the panelPart. You would then also have just one onCompositeOpen event instead of onViewletOpen and onPanelOpen.

File: src/vs/workbench/browser/parts/panel/panelPart.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+					timerEvent.stop();
+
+					// Fullfill promise with panel that is being opened
+					return panel;
+				});
+			});
+		});
+	}
+
+	private get activePanel(): IPanel {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/panel/panelPart.ts]: All of this is super nice and shared.

File: src/vs/workbench/browser/parts/sidebar/sidebarPart.ts [hunk 8]
```diff
@@ -182,363 +139,28 @@ export class SidebarPart extends Part implements IViewletService {
...
-
-		let keybinding: string = null;
-		let keys = this.keybindingService.lookupKeybindings(viewletId).map(k => this.keybindingService.getLabelFor(k));
-		if (keys && keys.length) {
-			keybinding = keys[0];
-		}
-
-		this.titleLabel.safeInnerHtml(viewletTitle);
-		this.titleLabel.title(keybinding ? nls.localize('viewletTitleTooltip', "{0} ({1})", viewletTitle, keybinding) : viewletTitle);
+	private createViewlet(id: string, isActive?: boolean): TPromise<Viewlet> {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/sidebar/sidebarPart.ts]: Similar to panelPart, I think more can be pushed down to compositePart.

File: src/vs/workbench/common/events.ts [hunk 9]
```diff
@@ -229,15 +234,15 @@ export class CommandEvent extends Event {
...
- * Viewlet events are emitted when a viewlet opens or closes in the sidebar.
+ * Composite events are emitted when a composite opens or closes in the sidebar or panel.
  */
-export class ViewletEvent extends Event {
-	public viewletId: string;
+export class CompositeEvent extends Event {
+	public compositeId: string;
 
-	constructor(viewletId: string, originalEvent?: any) {
+	constructor(compositeId: string, originalEvent?: any) {
```
Comment (bpasero) [inline src/vs/workbench/common/events.ts]: @isidorn I think here we should also send around the actual instance of Composite that was opened or closed because then one can simply do an instance of check to find out if the composite was a viewlet or panel.

File: src/vs/workbench/common/constants.ts [hunk 10]
```diff
@@ -16,6 +16,7 @@ export const Identifiers = {
 	WORKBENCH_CONTAINER: 'workbench.main.container',
 	ACTIVITYBAR_PART: 'workbench.parts.activitybar',
 	SIDEBAR_PART: 'workbench.parts.sidebar',
+	PANEL_PART: 'workbench.parts.panelpart',
```
Comment (bpasero) [inline src/vs/workbench/common/constants.ts]: @isidorn suggest to call it just "workbench.parts.panel"

File: src/vs/workbench/browser/workbench.ts [hunk 11]
```diff
@@ -79,6 +82,7 @@ export class Workbench implements IPartService {
 
 	private static sidebarPositionSettingKey = 'workbench.sidebar.position';
 	private static sidebarHiddenSettingKey = 'workbench.sidebar.hidden';
+	private static panelPartHiddenSettingKey = 'workbench.panelPart.hidden';
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: @isidorn suggest to call it just "workbench.panel.hidden"

File: src/vs/workbench/browser/workbench.ts [hunk 12]
```diff
@@ -112,6 +117,7 @@ export class Workbench implements IPartService {
 	private creationPromiseComplete: ValueCallback;
 	private sideBarHidden: boolean;
 	private sideBarPosition: Position;
+	private panelPartHidden: boolean;
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: => panelHidden

File: src/vs/workbench/browser/workbench.ts [hunk 13]
```diff
@@ -499,6 +533,42 @@ export class Workbench implements IPartService {
 		this.storageService.store(Workbench.sidebarHiddenSettingKey, hidden ? 'true' : 'false', StorageScope.WORKSPACE);
 	}
 
+	public isPanelPartHidden(): boolean {
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: @isidorn => isPanelHidden

File: src/vs/workbench/browser/workbench.ts [hunk 14]
```diff
@@ -499,6 +533,42 @@ export class Workbench implements IPartService {
...
+		return this.panelPartHidden;
+	}
+
+	public setPanelPartHidden(hidden: boolean, skipLayout?:boolean): void {
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: @isidorn => setPanelHidden

File: src/vs/workbench/browser/layout.ts [hunk 15]
```diff
@@ -18,7 +18,10 @@ import {IContextViewService} from 'vs/platform/contextview/browser/contextView';
 import {IEventService} from 'vs/platform/event/common/event';
 
 const DEFAULT_MIN_PART_WIDTH = 170;
+const DEFAULT_MIN_PANEL_PART_HEIGHT = 170;
 const HIDE_SIDEBAR_WIDTH_THRESHOLD = 50;
+const HIDE_PANEL_HEIGHT_THRESHOLD = 50;
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @isidorn <3

File: src/vs/workbench/browser/layout.ts [hunk 16]
```diff
@@ -48,33 +52,43 @@ interface ComputedStyles {
 /**
  * The workbench layout is responsible to lay out all parts that make the Monaco Workbench.
  */
-export class WorkbenchLayout implements IVerticalSashLayoutProvider {
+export class WorkbenchLayout implements IVerticalSashLayoutProvider, IHorizontalSashLayoutProvider {
 
-	private static sashWidthSettingsKey = 'workbench.sidebar.width';
+	private static sashXWidthSettingsKey = 'workbench.sidebar.width';
+	private static sashYHeightSettingsKey = 'workbench.panelpart.height';
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: => workbench.panel.height

File: src/vs/workbench/browser/layout.ts [hunk 17]
```diff
@@ -48,33 +52,43 @@ interface ComputedStyles {
...
 	private computedStyles: ComputedStyles;
-	private editorHeight: number;
 	private workbenchSize: Dimension;
-	private sash: Sash;
+	private sashX: Sash;
+	private sashY: Sash;
 	private startSidebarWidth: number;
 	private sidebarWidth: number;
+	private sidebarHeight: number;
+	private startPanelPartHeight: number;
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @isidorn again easier to read the thing as "panel" unless you actually talk about the panelPart class

File: src/vs/workbench/browser/layout.ts (hunk 16 above)
Comment (isidorn) [inline src/vs/workbench/browser/layout.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/layout.ts (hunk 17 above)
Comment (isidorn) [inline src/vs/workbench/browser/layout.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/compositePart.ts (hunk 3 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/compositePart.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/compositePart.ts (hunk 5 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/compositePart.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/workbench.ts (hunk 11 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/workbench.ts (hunk 12 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/workbench.ts (hunk 13 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/workbench.ts (hunk 14 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark:

File: src/vs/workbench/common/constants.ts (hunk 10 above)
Comment (isidorn) [inline src/vs/workbench/common/constants.ts]: :heavy_check_mark:

File: src/vs/workbench/common/events.ts (hunk 9 above)
Comment (isidorn) [inline src/vs/workbench/common/events.ts]: Just tried this and the only issue is that sidebarPart.ts:90 sends a CompositeEvent before he has the actual composite, he just has the id. I can send the event later, on line 108 but not sure if too late

File: src/vs/workbench/browser/parts/compositePart.ts (hunk 2 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/compositePart.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/compositePart.ts (hunk 4 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/compositePart.ts]: :heavy_check_mark:

File: src/vs/workbench/common/events.ts (hunk 9 above)
Comment (bpasero) [inline src/vs/workbench/common/events.ts]: Got it, I would not change it then. We can still tweak this later if someone needs to know the difference.

File: src/vs/workbench/browser/parts/activitybar/activitybarPart.ts (hunk 1 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/panel/panelPart.ts (hunk 6 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/panel/panelPart.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/sidebar/sidebarPart.ts (hunk 8 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/sidebar/sidebarPart.ts]: :heavy_check_mark:

File: src/vs/workbench/common/events.ts (hunk 9 above)
Comment (isidorn) [inline src/vs/workbench/common/events.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/composite.ts [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+/**
+ * Internal composite events to communicate with composite container.
+ */
+export const EventType = {
+	INTERNAL_COMPOSITE_TITLE_AREA_UPDATE: 'internalCompositeTitleAreaUpdate'
+};
+
+/**
+ * Composites are layed out in the sidebar part of the workbench. Only one composite can be open
```
Comment (bpasero) [inline src/vs/workbench/browser/composite.ts]: @isidorn would rephrase, it is not just sidebar

File: src/vs/workbench/browser/layout.ts (hunk 15 above)
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @isidorn is this used at all?

Comment (isidorn) [inline src/vs/workbench/browser/layout.ts]: @bpasero no. I wanted to have the nice action you have in sidebar to hide it when this threshold is reached but not sure if we want it since bringing it back would not be as easy as in the sidebar. Or you think that whole behavior makes sense for panel as well?

File: src/vs/workbench/browser/panel.ts [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+	constructor(
+		id: string,
+		name: string,
+		@IPartService private partService: IPartService
+	) {
+		super(id, name, 'close-editor-action');
+	}
+
+	public run(): Promise {
+		this.partService.setPanelHidden(false);
```
Comment (bpasero) [inline src/vs/workbench/browser/panel.ts]: I guess this should be "true" for closing?

File: src/vs/workbench/browser/parts/panel/media/panelPart.css [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+.monaco-workbench > .panel > .content > .panel {
+	height: 100%;
+}
+
+.monaco-workbench > .part.panel > .title {
+	display: block;
+	border-top: 1px solid #007ACC
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/panel/media/panelPart.css]: @isidorn is this value not dynamic depending on debug, no workspace or workspace? I would check what the statusbar does to set this color.

File: src/vs/workbench/browser/panel.ts [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+	public getDefaultPanelId(): string {
+		return this.defaultPanelId;
+	}
+}
+
+export const Extensions = {
+	Panels: 'workbench.contributions.panels'
+};
+
+export class ClosePanelAction extends Action {
```
Comment (bpasero) [inline src/vs/workbench/browser/panel.ts]: @isidorn Suggest to move these actions into panelPart. You can see that sidebarPart is also contributing an action!

File: src/vs/workbench/browser/workbench.ts [hunk 22]
```diff
@@ -208,6 +214,11 @@ export class Workbench implements IPartService {
 				viewletAndEditorPromises.push(this.sidebarPart.openViewlet(viewletId, false).then(() => viewletTimerEvent.stop()));
 			}
 
+			if (!this.panelHidden) {
+				const panelId = this.storageService.get(PanelPart.activePanelSettingsKey, StorageScope.WORKSPACE);
+				viewletAndEditorPromises.push(this.panelPart.openPanel(panelId, false));
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: @isidorn viewletAndEditorPromises => createCompositesPromises or so :)

File: src/vs/workbench/browser/workbench.ts [hunk 23]
```diff
@@ -391,11 +418,17 @@ export class Workbench implements IPartService {
 			this.sideBarHidden = true; // we hide sidebar in single-file-mode
 		}
 
-		let registry = (<IViewletRegistry>Registry.as(ViewletExtensions.Viewlets));
+		let registry = (<ViewletRegistry>Registry.as(ViewletExtensions.Viewlets));
 		if (!registry.getDefaultViewletId()) {
 			this.sideBarHidden = true; // can only hide sidebar if we dont have a default viewlet id
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: @isidorn for the sake of completness, the panel should also be hidden if there is no default panel

File: src/vs/workbench/browser/layout.ts (hunk 15 above)
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @isidorn not sure it would be harder, as a user you can drag from above the status bar to restore it. I think it would be consistent with how the sidebar behaves

File: src/vs/workbench/browser/composite.ts (hunk 18 above)
Comment (isidorn) [inline src/vs/workbench/browser/composite.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/panel/media/panelPart.css (hunk 20 above)
Comment (isidorn) [inline src/vs/workbench/browser/parts/panel/media/panelPart.css]: @bpasero status bar does the same thing, defines a default, and then clients like debug set it to orange. I do the same here.

File: src/vs/workbench/browser/workbench.ts (hunk 22 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark: 
I use compositeAndEditorPromises, since some editor promises get appended at the end and createCompositesPromises is not got due to this

File: src/vs/workbench/browser/workbench.ts (hunk 23 above)
Comment (isidorn) [inline src/vs/workbench/browser/workbench.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/parts/panel/media/panelPart.css (hunk 20 above)
Comment (bpasero) [inline src/vs/workbench/browser/parts/panel/media/panelPart.css]: Ack

File: src/vs/workbench/browser/workbench.ts (hunk 22 above)
Comment (bpasero) [inline src/vs/workbench/browser/workbench.ts]: Makes sense

File: src/vs/workbench/browser/panel.ts (hunk 21 above)
Comment (isidorn) [inline src/vs/workbench/browser/panel.ts]: :heavy_check_mark:

File: src/vs/workbench/browser/panel.ts (hunk 19 above)
Comment (isidorn) [inline src/vs/workbench/browser/panel.ts]: :heavy_check_mark:
